Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device.cpp

```diff
@@ -2,47 +2,137 @@
 // Licensed under GPLv2+
 // Refer to the license.txt file included.
 
-#include "Core/IPC_HLE/WII_IPC_HLE.h"
+#include <algorithm>
+
 #include "Common/StringUtil.h"
 #include "Core/HW/Memmap.h"
 #include "Core/HW/SystemTimers.h"
+#include "Core/IPC_HLE/WII_IPC_HLE.h"
 #include "Core/IPC_HLE/WII_IPC_HLE_Device.h"
 
-SIOCtlVBuffer::SIOCtlVBuffer(const u32 address) : m_Address(address)
+IOSResourceRequest::IOSResourceRequest(const u32 command_address) : address(command_address)
+{
+  command = static_cast<IPCCommandType>(Memory::Read_U32(command_address));  // 0x00
+  return_value = Memory::Read_U32(command_address + 4);                      // 0x04
+  fd = Memory::Read_U32(command_address + 8);                                // 0x08
+}
+
+void IOSResourceRequest::SetReturnValue(const s32 new_return_value)
+{
+  return_value = new_return_value;
+  Memory::Write_U32(return_value, address + 4);
+}
+
+// Write out the IPC struct from command_address to num_commands numbers of 4 byte commands.
+void IOSResourceRequest::DumpCommands(size_t num_commands, LogTypes::LOG_TYPE log_type,
+                                      LogTypes::LOG_LEVELS verbosity) const
+{
+  GENERIC_LOG(log_type, verbosity, "Dump of request 0x%08x (fd %d)", address, fd);
+  for (u32 i = 0; i < num_commands; i++)
```

Comment:
Since `num_commands` is a size_t, `i` should likely be one as well. printf specifier below would need to be changed to `%zu` in that case.